parla/parray/core.py

```diff
@@ -0,0 +1,458 @@
+import numpy
+import cupy
+
+from parla.cpu_impl import cpu
+from parla.cuda import gpu
+from parla.tasks import get_current_devices
+from parla.device import Device
+
+class PArray:
+    """Multi-dimensional array on a CPU or CUDA device.
+
+    This class is a wrapper around :class:`numpy.ndarray` and :class:`cupy.ndarray`,
+    It is used to support Parla sheduler optimization and automatic data movement
+
+    Args:
+        array: :class:`cupy.ndarray` or :class:`numpy.array` object
+    """
+    def __init__(self, array=None) -> None:
+        self.array = array
+
+    @property
+    def _on_gpu(self) -> bool:
+        """
+        True if the array is on GPU
+        """
+        return isinstance(self.array, cupy.ndarray)
+
+    @property
+    def _index(self) -> int:
+        """
+        Index of Current Device
+        -1 if on CPU
+        """
+        if self._on_gpu:
+            return self.array.device
+        return -1
+
+    @staticmethod
+    def _get_current_device() -> Device:
+        """
+        Get current device from task environment.
+        """
+        return get_current_devices()[0]
+
+    def _auto_move(self) -> None:
+        """
+        Automatically move array to current device.
+        """
+        device = PArray._get_current_device()
+        if device.architecture == gpu:
+            if self._index == device.index:
+                return
+            self._to_current_gpu()
+        elif device.architecture == cpu:
+            self._to_cpu()
+
+    def _to_current_gpu(self) -> None:
+        """
+        Move the array to current GPU, do nothing if already on the device.
+        """
+        self.array = cupy.asarray(self.array)  # asarray by default copy to current device
+
+    def _to_gpu(self, index: int) -> None:
```

Comment:
Do we need this function in the future?

Reply:
I meant you pointed me out that moving to the current gpu does not need id specification and i agreed with you :-)

Reply:
There is a "_to_current_gpu()", which is by default "_to_gpu(current_id)"
I leave this method since it may be used in the future to move data into a specific device rather than current one.